# Step 2: データ群の扱いを理解しよう。
- 2-1. リスト/配列の基本
- 2-2. 繰り返し処理(for文)
- 2-3. 組み込み関数

# Step 2 の学習目標
- 複数の値（観測データ）を **1つの変数にまとめて管理** できるようになる。  
- `for` 文を使ってデータを **1つずつ取り出して処理** できるようになる。  
- 組み込み関数 `sum()` と `len()` を使って **平均値を計算** できるようになる。  
- データ群の扱いを通して、**プログラムで集計する感覚** を身につける。

# 2-1. リスト/配列の基本

- Pythonでは、複数の値をまとめて扱うには **リスト** を使う。
- 角かっこ `[]` で囲み、カンマで区切る。

In [83]:
# サンプル：1980年8月1日〜7日までの日平均気温
temps_tokyo = [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
print("temps_tokyo のデータ数は", len(temps_tokyo))
print(temps_tokyo)

temps_tokyo のデータ数は 7
[21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]


- インデックス（番号）で特定の要素にアクセスできる。
- ⚠️ リストの先頭は1番目ではなく0番目のインデックスが割り当てられていることに注意。

In [84]:
print(temps_tokyo[0])  # 最初の要素（27.4）
print(temps_tokyo[6]) # 最後の要素（28.2）
print(temps_tokyo[-1]) # 最後の要素（28.2）

21.6
22.2
22.2


- リストの要素は置き換え、追加、削除が可能である。tempsは最初の時点ではtemps_tokyoと同じリストが代入されている。その後の操作でどのようにリストの要素や構造が変化するか確認してみよう。

In [85]:
temps = [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
print("--- 初期 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)
temps[3] = 28.1
print("--- 置換後 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)

temps.append(29.0) # list.append(x) は list の末尾に x を追加
print("--- 末尾追加後 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)

temps.insert(1, 27.8) # list.insert(i, x) は list の i 番目の位置に x を挿入
print("--- 挿入後 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)

del temps[0] # del list[i] は list の i 番目の位置の要素を削除
print("--- 削除後 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)

--- 初期 ---
      temps: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
temps_tokyo: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
--- 置換後 ---
      temps: [21.6, 20.5, 19.5, 28.1, 21.8, 21.4, 22.2]
temps_tokyo: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
--- 末尾追加後 ---
      temps: [21.6, 20.5, 19.5, 28.1, 21.8, 21.4, 22.2, 29.0]
temps_tokyo: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
--- 挿入後 ---
      temps: [21.6, 27.8, 20.5, 19.5, 28.1, 21.8, 21.4, 22.2, 29.0]
temps_tokyo: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
--- 削除後 ---
      temps: [27.8, 20.5, 19.5, 28.1, 21.8, 21.4, 22.2, 29.0]
temps_tokyo: [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]


- 以下では temps に temps_tokyo を代入し、temps[0]を置換した。

In [87]:
temps = temps_tokyo
print("--- 初期 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)
temps[0] = 0.0
print("--- 置換後 ---")
print("      temps:",temps)
print("temps_tokyo:",temps_tokyo)

--- 初期 ---
      temps: [0.0, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
temps_tokyo: [0.0, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
--- 置換後 ---
      temps: [0.0, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
temps_tokyo: [0.0, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]


- 上の例では、temps = temps_tokyo と書いた時点で、新しいリストを作っているわけではなく、
temps_tokyo という同じリストを指す別名（別のラベル）を作っただけになる。
そのため、temps[0] = 0.0 と変更すると、同じリストを共有している temps_tokyo の中身も一緒に変わる。
- つまり、**2つの変数が同じメモリ上のリストを参照している状態** となることに注意が必要。
- listがベクトルである場合は list.copy()で独立したリストを作成できる。

In [ ]:
temps = temps_tokyo.copy()
temps[0] = 0.0

print("--- コピー後 ---")
print("      temps:", temps)
print("temps_tokyo:", temps_tokyo)

> 💡 **Tips:**  
> Pythonでは、変数は「値を入れる箱」ではなく「値がある場所へのラベル」である。  
> たとえば
>
> ```python
> x = 10  
> y = x
> ```
> 
> とすると、x と y は同じ「10」という値を指す。
> 
> コンピュータは、計算の途中で扱うデータを メモリ（記憶領域） に一時的に保存する。
> 私たちが Python で x = 10 と書いたとき、コンピュータの中では次のようなことが起きている：
>
> 1. まず、10という値を保存するための領域がメモリ上に確保される。
> 2. その領域に「x という名前（ラベル）」が貼られる。
> 3. 以後、x と書けば、その領域の中の値（10）を読み書きできる。
>
> Fortran, C, Javaなどでは、まずメモリ上に変数を確保し（変数宣言）、その後に値を代入するが、Pythonの場合は値を作ってから、それに名前（ラベル）を貼る作業をしている。

- sorted(), sort()関数を使うことでリストを昇順、降順で並べ替えることができる。後に述べる組み込み関数を用いない場合、並べ替えによって最小値や最大値を見つけることができる。
- sorted()は実行時に一時的にリストの順序を入れ替えるのに対し、とsort()は永続的にリストの順序を入れ替える。

In [101]:
temps_tokyo = [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]

print("--- sorted(list) (昇順) による並べ替え ---\n") #最後の"\n"は改行コード
print("sorted temps_tokyo: ", sorted(temps_tokyo))
print("       temps_tokyo: ",temps_tokyo, "\n")

print("--- sorted(list, reverse=True) (降順) による並べ替え ---\n")
print("sorted temps_tokyo: ", sorted(temps_tokyo, reverse=True))
print("       temps_tokyo: ",temps_tokyo, "\n")

print("--- list.sort() (昇順) による並べ替え ---\n")
temps_tokyo.sort()
print("       temps_tokyo: ",temps_tokyo, "\n")

print("--- list.sort(reverse=True) (降順) による並べ替え ---\n")
temps_tokyo.sort(reverse=True)
print("       temps_tokyo: ",temps_tokyo)

--- sorted(list) (昇順) による並べ替え ---

sorted temps_tokyo:  [19.5, 20.5, 20.5, 21.4, 21.6, 21.8, 22.2]
       temps_tokyo:  [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2] 

--- sorted(list, reverse=True) (降順) による並べ替え ---

sorted temps_tokyo:  [22.2, 21.8, 21.6, 21.4, 20.5, 20.5, 19.5]
       temps_tokyo:  [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2] 

--- list.sort() (昇順) による並べ替え ---

       temps_tokyo:  [19.5, 20.5, 20.5, 21.4, 21.6, 21.8, 22.2] 

--- list.sort(reverse=True) (降順) による並べ替え ---

       temps_tokyo:  [22.2, 21.8, 21.6, 21.4, 20.5, 20.5, 19.5]


- list.reverse()はリストを逆順で永続的に入れ替える。

In [107]:
temps_tokyo = [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]
temps_tokyo.reverse()
print("--- list.reverse() (逆順) による並べ替え ---\n") #最後の"\n"は改行コード
print("reversed temps_tokyo: ", temps_tokyo)

--- list.reverse() (逆順) による並べ替え ---

reversed temps_tokyo:  [22.2, 21.4, 21.8, 20.5, 19.5, 20.5, 21.6]


# 2-2. 繰り返し処理(for文)

## Pythonのfor文の構文

- 複数のデータを1つずつ取り出して処理するときには、**for文（繰り返し処理）** を使う。
- 所定回数の繰り返しには、 `for` と `range(...)` を組み合わせる。

---

```python
for i in range(N):
    # i は 0, 1, 2, ..., N-1 と変化
    # ここに繰り返したい処理
```
---
- for文で繰り返したい処理は、必ずインデント（字下げ）して記載する。インデントから始まる行が繰り返し処理の中身であることを示す。

- range(5)は 0, 1, 2, 3, 4 の整数の並びを生成する。以下では i が 0 - 4 まで繰り返し表示される。

In [9]:
for i in range(5):
    print(i)

0
1
2
3
4


- `range()` 関数は、整数の等差数列（arithmetic sequence）を生成する関数である。  
- 形式的には次のように表せる：

  $$
  \text{range}(a,\, b,\, s) = \{\, a,\; a+s,\; a+2s,\; \ldots,\; a+ks \mid a+ks < b \,\}
  $$
  
  （ただし $s > 0$ の場合）

  → 初項 $a$、公差$s$ の等差数列を作り、終端値$b$ は含まれない。

In [119]:
tmp = list(range(5))          # range(0, 5, 1) と同じ
print(tmp)                    # [0, 1, 2, 3, 4]

tmp = list(range(0, 11, 2))   # 初項0、公差2、終端11（11は含まれない）
print(tmp)                    # [0, 2, 4, 6, 8, 10]

[0, 1, 2, 3, 4]
[0, 2, 4, 6, 8, 10]


- 以下ではtemps_tokyoのデータ数で範囲指定し(range(7))、temps_tokyo[0]からtemps_tokyo[6]までを表示する。

In [35]:
for i in range(len(temps_tokyo)):
    print(temps_tokyo[i])

27.4
27.1
28.0
29.3
28.8
27.9
28.2


- 以下の構文では、リストの全要素を順に変数に代入し、繰り返し処理を行う。
---
```python
for 変数 in データ群:
    実行したい処理
```
---

- 先ほどはtemps_tokyoの位置をiで指定し繰り返し表示したが、以下ではtemps_tokyoリストからtemp変数に繰り返し代入し、都度tempを表示する。

In [138]:
temps_tokyo = [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2]

for tmp in temps_tokyo:
    print(tmp)

21.6
20.5
19.5
20.5
21.8
21.4
22.2


- for文の中では複数行の処理を実施することができる。for文で実施することを示すために、繰り返し実施する部分は全てインデントする。以下、インデント位置の異なるprint関数の挙動を確認してみよう。

In [137]:
sum = 0
print(f"合計: {sum}")
for temp in temps_tokyo:
    print(temp)
    sum = sum + temp
    print(f"小計: {sum}")
print(f"合計: {sum}")

合計: 0
21.6
小計: 21.6
20.5
小計: 42.1
19.5
小計: 61.6
20.5
小計: 82.1
21.8
小計: 103.89999999999999
21.4
小計: 125.29999999999998
22.2
小計: 147.49999999999997
合計: 147.49999999999997


- print(f"小計: {sum}")は インデントされているため、for 文の繰り返しごとに実行される。  
    → sum（最初は0）に、temps_tokyo の各値を順に加算する過程が表示される。  

- 一方、print(f"合計: {sum}")はインデントされていないため、for 文の繰り返しの外にあり、ループがすべて終了したあとに一度だけ実行される。つまり、sum は次のように更新されていく：

$$
\text{sum} = \sum_{i=1}^{n} \text{list}_i
$$

  ここで、$\text{list}_i$ は list の$i$番目の要素、$n$はリストの要素数。  
- したがって、最終的な sum はリスト全体の総和になります。  

> 💡 **Tips:**  
> Pythonで小数の計算を行うと、結果が `147.49999999999997` のように小数点以下に微妙なズレが生じることがある。  
> これはPythonの計算ミスではなく、コンピュータ内部での数値表現の仕組みが原因。  
>
> コンピュータは数値を 2進数（二進法）で扱うが、10進数の小数（たとえば 0.1 や 21.6）は、  
> 2進数では無限に続く小数になり、正確に表現することができない。  
> そのため、Python（やC、Javaなどほとんどの言語）では、これらの小数を「近似値」として記録する。  
>
> たとえば:
> ```python
> print(0.1 + 0.2)
> ```
> 出力：
> ```
> 0.30000000000000004
> ```
> のように、わずかな誤差が発生ずる。 

# 2-3. 組み込み関数

- 1週間分の日平均気温データから、for文を用いて1週間の平均気温を計算する方法を考えてみよう。
- 先ほどのように、for文の中で気温データの総和を計算し、最後に日数で割ることで平均を得ることができる。
$$
\text{ave.} = \frac{1}{n}\sum_{i=1}^{n} \text{list}_i
$$

In [139]:
sum = 0
for tmp in temps_tokyo:
    sum = sum + tmp
print(f"総和: {sum}")
print(f"平均: {sum/len(temps_tokyo):.2f}")

総和: 147.49999999999997
平均: 21.07


> 💡 **Tips:**  
> 　上記の計算では1週間(7日間)の平均を計算するため、sum/7として平均を計算しても良い。
> ただし、リストの長さを使って sum / len(temps_tokyo) のように書いておくと、データの数が変わっても自動的に正しい平均が計算することができる。
> 　たとえば、 temps_tokyo に8日分のデータを追加しても、 len(temps_tokyo) は自動的に8を返すので、コードを直す必要はない。  
> こうした「データが増減しても書き換え不要で動くコード」は、再利用しやすく、ミスも減らせるという点でとても大切。

- 実は、sum(list)関数というlistの総和を返す関数もある。自分でfor文を用いて計算した結果と比較してみよう。

In [128]:
sum(temps_tokyo)

TypeError: 'float' object is not callable

- 上記を実行するとエラーが出るはずだ。このエラーは一個上のセルでPythonにもともと存在していた組み込み関数sumに数値を上書きしてしまったことが原因である。このように、気付かないうちに変数や関数に上書きしてしまってエラーが生じることはしばしば発生する。こうしたときは、一度変数や関数をprintで表示してみるのが定石。

In [129]:
print(sum)

196.7


- 上書きされてしまったsumを解放し、組み込み関数に戻すためには、以下のように削除を行う。

In [130]:
del sum

- もう一度printで表示してみよう。<built-in function sum> (組み込み関数 sum) が表示されるはず。もう一度sum関数による総和の計算を試してみよう。

In [131]:
print(sum)
sum(temps_tokyo)

<built-in function sum>


147.5

- 以下にいくつかの組み込み関数を紹介し、その組み込み関数の働きをfor文で実装する例を紹介する。

| 関数         | 説明        |
| ---------- | --------- |
| `sum(list)` | すべての値の合計  |
| `len(list)` | データ数（要素数） |
| `max(list)` | 最大値       |
| `min(list)` | 最小値       |


- 以下では組み込み関数 max(list) を用いてリストの最大値を求める。

In [141]:
maxv = max(temps_tokyo)
print(f"最大値: {maxv:.2f}")

最大値: 22.20


- for 文を使って自分で最大値を見つける仕組みを実装し、max() がどのように動いているのかを理解してみよう。

In [142]:
maxv = temps_tokyo[0]           # まず最初の値を「仮の最大値」として設定
for temp in temps_tokyo:        # リストの中身を1つずつ取り出して比較
    if maxv < temp:             # 今の最大値より大きければ
        maxv = temp             # 新しい値を最大値として更新
print(f"最大値: {maxv:.2f}")

最大値: 22.20


# 2-4. DataFrame

- 複数のリスト(列)を束ねると表になる。表データは Pandasライブラリ が提供する DataFrame 形式で扱うことが多い。※ DataFrame についてはまた後ほど説明。
- Pandasライブラリはデータ解析機能を提供するPythonのデータ解析ライブラリである。

In [126]:
import pandas as pd

date_index = [1, 2, 3, 4, 5, 6, 7]
temps_tokyo_1980 = pd.Series([21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2], index = date_index)
print(temps_tokyo_1980,'\n')

date_index = [1, 2, 3, 4, 5, 6, 7]
temps_tokyo_1990 = pd.Series([26.6, 28.2, 28.4, 27.7, 18.5, 28.8, 29.6], index = date_index)
print(temps_tokyo_1990,'\n')

df = pd.DataFrame({'1980':temps_tokyo_1980, '1990':temps_tokyo_1990})

print(df)

# 列ごとのリストを辞書としてまとめる
data = {
    "date": [1, 2, 3, 4, 5, 6, 7],
    "temps_tokyo_1980": [21.6, 20.5, 19.5, 20.5, 21.8, 21.4, 22.2],
    "temps_tokyo_1990": [26.6, 28.2, 28.4, 27.7, 18.5, 28.8, 29.6],
} #ここに説明を入れたい

df = pd.DataFrame(data) #ここに説明を入れたい
print(df)

1    21.6
2    20.5
3    19.5
4    20.5
5    21.8
6    21.4
7    22.2
dtype: float64 

1    26.6
2    28.2
3    28.4
4    27.7
5    18.5
6    28.8
7    29.6
dtype: float64 

   1980  1990
1  21.6  26.6
2  20.5  28.2
3  19.5  28.4
4  20.5  27.7
5  21.8  18.5
6  21.4  28.8
7  22.2  29.6
   date  temps_tokyo_1980  temps_tokyo_1990
0     1              21.6              26.6
1     2              20.5              28.2
2     3              19.5              28.4
3     4              20.5              27.7
4     5              21.8              18.5
5     6              21.4              28.8
6     7              22.2              29.6


- 【辞書の説明を入れたい】

- DataFrameの列名、行数、全体の構造はdf.xxxで取得することができる。【ここももう少し詳しい説明がいるかも。】

In [50]:
# DataFrame の構造要素
print('\n--- 行と列の情報 ---')
print('列名 columns:', df.columns.tolist())
print('行番号 index:', df.index.tolist())
print('形状 shape:', df.shape)


--- 行と列の情報 ---
列名 columns: ['date', 'temps_tokyo_1980', 'temps_tokyo_1990']
行番号 index: [0, 1, 2, 3, 4, 5, 6]
形状 shape: (7, 3)


- DataFrameの中身を取り出すときには、locもしくはilocで指定する方法がある。
    - loc  → ラベル（index名や列名）で指定する。
    - iloc → 数値インデックス（位置）で指定する。
- 【コロンの説明、行と列をどのように指定しているのかの説明が要る】

In [57]:
df.iloc[:,1]

0    21.6
1    20.5
2    19.5
3    20.5
4    21.8
5    21.4
6    22.2
Name: temps_tokyo_1980, dtype: float64

In [58]:
df.loc[:,'temps_tokyo_1980']

0    21.6
1    20.5
2    19.5
3    20.5
4    21.8
5    21.4
6    22.2
Name: temps_tokyo_1980, dtype: float64

- 1980年から2020年の東京の8月各日付の気温データを列でならべた場合、40年 x 31日の表データになる。
- それぞれの列について、どうやったら手際良く繰り返し平均を求めることができるだろう？
- DataFrameとはExcelの表のような「行×列」構造をもつ2次元データである。
- Pandasライブラリでは1列データ（1列ベクトル）を Series型として扱う。Seriesを複数まとめたものがDataFrame型である。以下ではSeriesにそれぞれ辞書（列名）を与えた7行 x 3列のDataFrameがdfに代入される。
- 【2重ループの例題が欲しい】

# Step 2の練習問題
- 【あればあるだけ良い。環境に関わる問題だと良い。】